ipv-core/src/main/java/com/ibm/whc/deid/providers/masking/AbstractMaskingProvider.java

```diff
@@ -110,22 +112,39 @@ public void maskIdentifierBatch(List<MaskingActionInputIdentifier> identifiers)
   }
 
   protected final void putField(MaskingActionInputIdentifier i, String value) {
+    // TextNode does not expect null as its value, so use NullNode.
+    // TextNode typically works, but as of Jackson 2.11.4, some methods like hashCode() get NPE.
+    JsonNode newNode = value == null ? NullNode.getInstance() : new TextNode(value);
     if (i.getParent().isObject()) {
-      TextNode newNode = new TextNode(value);
-      i.setParent(((ObjectNode) i.getParent()).set(i.getPath(), newNode));
+      ((ObjectNode) i.getParent()).set(i.getPath(), newNode);
       i.setCurrentNode(newNode);
     } else if (i.getParent().isArray()) {
-      ArrayNode aNode = (ArrayNode) i.getParent();
-      int indexOfResult = -1;
-      int size = aNode.size();
-      for (int currentIndex = 0; currentIndex < size; currentIndex++) {
-        if (aNode.get(currentIndex) == i.getNode()) {
-          indexOfResult = currentIndex;
-          break;
-        }
+      // Do not use object identity to find the array member.
+      // Jackson flyweights certain node objects so the same JsonNode
+      // can appear multiple times within a document tree.
+      String path = i.getPath();
+      int index = path.lastIndexOf('[');
+      if (index < 0) { // safety check - should not occur
+        throw new RuntimeException("path to member of array node does not contain `[`");
+      }
+      int closeIndex = path.indexOf(']', index);
+      if (closeIndex < 0) { // safety check - should not occur
+        throw new RuntimeException("path to member of array node does not contain `]` after `[`");
+      }
+      int offset;
+      try {
+        offset = Integer.parseInt(path.substring(index + 1, closeIndex));
+      } catch (NumberFormatException e) {
+        // should not occur
+        throw new RuntimeException("offset in path cannot be converted to an integer");
+      }
+      try {
+        ((ArrayNode) i.getParent()).set(offset, newNode);
+      } catch (IndexOutOfBoundsException e) {
+        // should not occur
+        throw new RuntimeException("offset in path was " + offset + " but ArrayNode has only "
+            + ((ArrayNode) i.getParent()).size() + " members");

```

Comment:
Seems to have many branches(if/try), and I don't see unit tests added for AbstractMaskingProvider.  We can consider adding unit tests for the future.

Reply:
Most of these checks are not actually necessary.  We create the strings its examining internally in the code, so we would expect them to be formatted correctly.  Its not direct data from the user.  All of our existing unit tests that to end-to-end masking flow through this method to update the JSON document.